README.txt
-----------------------------------------------------------------[ 0. Welcome ]-

Thank you for downloading dhex 0.69. It is a hex editor with a diff mode.
Since it is based on ncurses, it runs in any console.
It makes heavy use of colors, but it is themeable to run on monochrome-displays
as well.

If you have any questions, or found any bugs, please do not hesitate to contact
me under dettus@dettus.net. Be sure to include "DHEX" somewhere in the subject.


Thomas Dettbarn, 30-Dec-2010

0.60: Initial release
0.61: Some code cleanup, rudementary manual correlation is possible through 
      commandline parameters
0.62: Code cleanup, bug fixes, and as new features: bookmarks and ascii-search.
0.63: Man pages were written, searching is possible from the command line.
0.64: Naive correlation, a new default theme
0.65: dhex 0.64 crashed at startup on some systems.
0.66: searching for an ascii string with spaces works, some problems with the
      color themes were fixed
0.67: a "base address" functionality was added, which makes it easier to work
      with memory dumps.
0.69: this will be the final release of dhex 0.6x. I fixed a bug which caused
      the terminal to be broken. And added the functionality to start the 
      hexcal from the commandline (-x).

---------------------------------------------------------[ 1. Getting started ]-

Download the latest version from http://www.dettus.net/dhex/dhex_latest.tar.gz.
Unpack it,

% tar xvfz dhex_latest.tar.gz

then compile it

% make

This should produce an executable. It requires a file as a paramter to start

% ./dhex gpl.txt

When starting for the first time, dhex will attempt to create a .dhexrc file 
in your $HOME-directory. If you already have a .dhexrc from an older version of
dhex, it will convert it. 


The first screen you'll see will be the keyboard setup. (I was not satisfied 
with the way ncurses handled the function keys, so i created something new)
The keyboard setup will be written into the .dhexrc file.


You can edit the file, search for short strings in it, and scroll through it.
After pressing F10 (or whichever key you pressed instead of F10 during the 
keyboard setup), dhex will ask you whether or not to save the changes.


To invode the keyboard setup manually, just start dhex with the parameter -k

% ./dhex -k gpl.txt



---------------------------------------------------------------[ 2. Diff mode ]-


When you run dhex with two files, 

% ./dhex gpl.txt dhex

it will start in the diff mode. This is useful for file comparison. Any 
difference in the files will be highlighted. To jump from one difference to 
the next, press F3. To go to the previous one, press F4.


--------------------------------------------------------------------[ 3. Goto ]-

Pressing F1 will open up the "Goto" dialog. Here you have three possibilities:
Pressing = will set the cursor to the absolute address you're about to type in,
Pressing + will advance the cursors those many bytes,
Pressing - will go back those many bytes.

Press "Go" to set the cursor, or "Cancel" to quit the dialog.


Moreover, you can set up to ten bookmarks here. And save them for later use.



---------------------------------------------------------------[ 4. Searching ]-

When not in diff mode, pressing F2 will invoke the "Search" dialog. You can
search forward or backward. When selecting the "Writelog" option, you'll have
to provide a filename to write the search results to. 
Later, when you select the "Readlog" option, it'll read the cursor positions 
where to search from this file. Otherwise dhex will search the whole file.

When not selecting the "Writelog" option, the search dialog will immediately
close upon finding the first occuracnce, and set the cursor there.

Press F3 to go to the next occurace, or F4 to the previous one.

Searching is also possible from the command line: Simply apply the -sa, -sab, 
-sh, -shb parameters. To write into a searchlog, -w is the parameter to go. To 
read from it, -r can be used. For example

% ./dhex -oh affe -shb d00f dhex

sets the cursor to 0xaffe, and searches for the hex string "d0 0f" backwards 
from there.

% ./dhex -oh affe -shb d00f -w occurances.log dhex 

writes the results into a logfile called "occurances.log" and exits.

-----------------------------------------------------------------[ 5. Hexcalc ]-

Pressing F5 will open up the hexcalc dialog. This is a small 64-bit calculator.
Simply type in a value in one of the formats "decimal", "hexadecimal" or
"binary", chose an operation, press ENTER and see the results.
You can select any former result by going up a few lines with the cursor keys.



------------------------------------------------------------[ 6. config-files ]-

The default config file is $HOME/.dhexrc. Another config file can be chosen
with the -f option:

% ./dhex -f /etc/globaldhexrc gpl.txt

Basically, a config file is broken down into two sections: Colors and keyboard
conversion tab. The default config file on a MacBook looks like this:

DHEXCOLORSCHEME
#VERSION 0
#possible colors are: BLACK,GREEN,GREEN,YELLOW,BLUE,MAGENTA,CYAN,BLACK
#possible extra flags are: UNDERLINE,REVERSE,BLINK,DIM,BOLD

BRACKETS:       FG=BLACK,BG=BLACK,BOLD
HEXFIELD:       FG=WHITE,BG=BLACK
INPUT:          FG=BLACK,BG=WHITE
CURSOR:         FG=WHITE,BG=BLACK
TEXT:           FG=WHITE,BG=BLACK,BOLD
MENU_NORMAL:    FG=CYAN,BG=BLACK
MENU_HIGHLIGHT: FG=BLACK,BG=CYAN
MENU_HOTKEY:    FG=YELLOW,BG=BLACK,BOLD
MENU_HOTKEY_HI: FG=YELLOW,BG=CYAN,BOLD
FRAME:          FG=BLUE,BG=BLACK,BOLD
NORMAL_DIFF:    FG=YELLOW,BG=BLACK
CURSOR_DIFF:    FG=YELLOW,BG=WHITE,BOLD
HEADLINE:       FG=BLUE,BG=BLACK,BOLD
	
KEYF1:c2 a1 
KEYF2:e2 84 a2 
KEYF3:c2 a3 
KEYF4:c2 a2 
KEYF5:e2 88 9e 
KEYF6:c2 a7 
KEYF7:c2 b6 
KEYF8:e2 80 a2 
KEYF9:c2 aa 
KEYF10:c2 ba 
KEYESC:1b 
KEYBACKSPACE:08 
KEYDEL:1b 
KEYENTER:0a 
KEYTAB:09 
KEYUP:1b 5b 41 
KEYDOWN:1b 5b 42 
KEYRIGHT:1b 5b 43 
KEYLEFT:1b 5b 44 
KEYPGUP:02 
KEYPGDOWN:06 
KEYHOME:5e 
KEYEND:24 

(Actually, in this example ALT+1, ALT+2, ALT+3,... ALT+0 were used instead of 
F1..F10) You can edit it with any texteditor you like.



--------------------------------------[ 7. Correlating two files in diff mode ]-

Lets assume you would like to compare two files. One of those files starts with
garbage at the beginning:

  file1: GARBAGE0123456789ABCDEF
  file2:        0123456789ABCDEF

GARBAGE has 7 letters, so to skip over the first 7 bytes in the first file, just
apply the following paramters:

% dhex -o1d 7 file1 file2

-o sets the cursor offset. -o1 means that the first buffer is affected. -o1d 
gives the radix for the offset. it can either be -o1d for decimal, -o1b for
binary, -o1h for hexadecimal or -o1o for octal.

If the garbage is in the second file, just use

% dhex -o2d 7 file1 file2

Since version 0.64 there is also a way of doing this automatically: Pressing
F6 in the GUI will open up a dialog where you can select one of three
correlation algorithms. They are painfully slow, but they'll do the job.
Those three algorithms are:

- best match
  move the two files in such a way that as many bytes match as possible
- longest match
  move the two files in such a way that as many consecutive bytes match as
  possible
- minimum diff
  move the two files in such a way that as little bytes differ as possible. This
  algorithm requires an upper limit to improve its speed.

To invoke correlation from the command line, use one of these

% dhex -cb file1 file2  # best match
% dhex -cl file1 file2  # longest match
% dhex -cd 1000 file1 file2 # minimum difference, with the upper limit of 1000.
  


--------------------------------------------------------------[ 8. Disclaimer ]-

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


--------------------------------------------------------------[ 9. Known bugs ]-

- Searching has not been thourughly tested.
- Code is not clean (yet)
- Searching is only possible in normal mode, not diff mode
- Searching ignores unsaved changes.
- Some features are missing
- -k expects a file to be opened
- -f might overwrite ANY file.
- GOTO can jump way behind the end of the file
- some ways of exiting won't free all the allocated buffers


debian/changelog
dhex (0.69-1.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix cpu-load patch to enable nodelay() before keyboardsetup() and
    redisable it after (Closes: #940536) (LP: #1814478)

 -- nabijaczleweli <nabijaczleweli@gmail.com>  Fri, 20 Mar 2020 12:55:23 +0100

dhex (0.69-1) unstable; urgency=medium

  * New upstream version.
  * debian/control: drop Vcs fields.
  * Bump standards version to 4.2.1.
  * debian/copyright: updated years.
  * debian/install and manpages: added.
  * debian/rules: disabled install target.

 -- Gürkan Myczko <gurkan@phys.ethz.ch>  Tue, 22 Jan 2019 08:06:27 +0100

dhex (0.68-3) unstable; urgency=medium

  * Add myself to Uploaders.
  * Drop stripping, thanks Helmut Grohne. (Closes: #865731)
  * Fix busy waiting for input. (Closes: #714014)
  * Bump debhelper version to 11.
  * Bump standards version to 4.1.3.

 -- Gürkan Myczko <gurkan@phys.ethz.ch>  Wed, 14 Mar 2018 08:27:56 +0100

dhex (0.68-2) unstable; urgency=low

  * Upload to unstable because Wheezy is now stable. \o/

 -- Jonathan McCrohan <jmccrohan@gmail.com>  Mon, 06 May 2013 11:41:52 +0100

dhex (0.68-1) experimental; urgency=low

  * New upstream release
  * New maintainer (Closes: #691818)
    - Update debian/{control,copyright} accordingly.
  * Add debian/patches/dhex.1-man-typo to fix errors caught by linitian
  * Import packaging git repo to collab-maint
    - Add VCS-{Git,Browser} URLs
  * Bump S-V to 3.9.4
    - No changes needed
  * Fix Copyright Format 1.0 syntax error in debian/copyright
    - s/GNU GPL 2\+/GPL-2+/
  * Update package long description to aid searching
  * Thanks to Gregor Herrmann for reviewing and sponsoring this upload

 -- Jonathan McCrohan <jmccrohan@gmail.com>  Wed, 31 Oct 2012 22:38:35 +0000

dhex (0.67-1) unstable; urgency=low

  [ Gürkan Sengün ]
  * New upstream version. (Closes: #674782)
  * Bump standards version to 3.9.3.

  [ Axel Beckert ]
  * Enable hardening build flags
    + Bump debhelper compatibility to 9
    + Patch Makefile slightly by doing some s/=/+=/
  * Bump debian/copyright format to 1.0

 -- Gürkan Sengün <gurkan@phys.ethz.ch>  Fri, 22 Jun 2012 14:52:55 +0200

dhex (0.65-1) unstable; urgency=low

  * Initial release. (Closes: #613929)

 -- Gürkan Sengün <gurkan@phys.ethz.ch>  Thu, 27 Jan 2011 08:34:24 +0100


configfile.c
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include <string.h>
#include "configfile.h"
#include "machine_type.h"
#include "ncurses.h"
#include "output.h"
#include "input.h"

typedef	struct _tColorName
{
	char name[16];
	short len;
	short val;
	int attrs;
} tColorName;

typedef struct _tColorGroup
{
	char 	configname[16];
	short	 len;
	uicolors uicol;		
} tColorGroup;

#define	NUMCOLORGROUPS	14
const tColorGroup colorGroups[NUMCOLORGROUPS]={
				{"BRACKETS:",		 9,COLOR_BRACKETS},
				{"HEXFIELD:",		 9,COLOR_HEXFIELD},
				{"INPUT:",		 6,COLOR_INPUT},
				{"CURSOR:",		 7,COLOR_CURSOR},
				{"TEXT:",		 5,COLOR_TEXT},
				{"MENU_NORMAL:",	12,COLOR_MENUNORMAL},
				{"MENU_HIGHLIGHT:",	15,COLOR_MENUACTIVE},
				{"MENU_HOTKEY:",	12,COLOR_MENUHOTKEY},
				{"MENU_HOTKEY_HI:",	15,COLOR_MENUHOTKEYACTIVE},
				{"NORMAL_DIFF:",	12,COLOR_DIFF},
				{"CURSOR_DIFF:",	12,COLOR_CURSORDIFF},	
				{"HEADLINE:",		 9,COLOR_HEADLINE},
				{"HEADER:",		 7,COLOR_HEADER},
				{"FRAME:",		 6,COLOR_FRAME}
				};
#define	NUMCOLORNAMES	20
const tColorName colorNames[NUMCOLORNAMES]={
					{"BLACK",	5,COLOR_BLACK,0},
					{"RED",		3,COLOR_RED,0},
					{"GREEN",	5,COLOR_GREEN,0},
					{"BLUE",	4,COLOR_BLUE,0},
					{"PURPLE",	6,COLOR_MAGENTA,0},
					{"MAGENTA",	7,COLOR_MAGENTA,0},
					{"CYAN",	4,COLOR_CYAN,0},
					{"WHITE",	5,COLOR_WHITE,0},
					{"YELLOW",	6,COLOR_YELLOW,A_BOLD},
					{"BROWN",	5,COLOR_YELLOW,0},

					{"LIGHTBLACK",	10,COLOR_BLACK,A_BOLD},
					{"LIGHTRED",	 8,COLOR_RED,A_BOLD},
					{"LIGHTGREEN",	10,COLOR_GREEN,A_BOLD},
					{"LIGHTBLUE",	 9,COLOR_BLUE,A_BOLD},
					{"LIGHTPURPLE",	11,COLOR_MAGENTA,A_BOLD},
					{"LIGHTMAGENTA",12,COLOR_MAGENTA,A_BOLD},
					{"LIGHTCYAN",	 9,COLOR_CYAN,A_BOLD},
					{"LIGHTWHITE",	10,COLOR_WHITE,A_BOLD},
					{"LIGHTYELLOW",	11,COLOR_YELLOW,A_BOLD},
					{"LIGHTBROWN",	10,COLOR_YELLOW,A_BOLD}
				};

#define	NUMATTRNAMES	5
const tColorName attrNames[NUMATTRNAMES]={
					{"UNDERLINE",	 9,0,A_UNDERLINE},
					{"REVERSE",	 7,0,A_REVERSE},
					{"BLINK",	 5,0,A_BLINK},
					{"DIM",		 3,0,A_DIM},
					{"BOLD",	 4,0,A_BOLD}
				};

int getcolors(tOutput* output,char* line)
{
	int i;
	int l;
	int state=0;	// 0: find :, 1=find FG=, 2=find, , 3=find BG=, 4=find , ,5=find END, 6=nothing
	char token[6][64];
	int tokenidx=0;
	char c;
	short col1;
	short col2;
	int attrs1;
	int attrs2;
	int attrsret;
	short fg=7,bg=0;
	
	l=strlen(line);
	memset(token,0,sizeof(token));
	for (i=0;i<l;i++)
	{
		c=line[i];
		if (c=='#') state=6;
		if (c==' ' || c==9) c=0;	// ignore all the white spaces
		else {
			if (c>='a' && c<='z') c-=32;	// make the letters lower case
			if (state==0 && c==':') {
				token[state][tokenidx++]=c;	//apend the last char
				token[state][tokenidx]=0;	
				state=1;
				tokenidx=0;
			}
			else if (state==1 && c=='=') {state=2;tokenidx=0;}
			else if (state==2 && c==',') {state=3;tokenidx=0;}
			else if (state==3 && c=='=') {state=4;tokenidx=0;}
			else if (state==4 && c==',') {state=5;tokenidx=0;}
			else if (state==5 && c<32) {state=6;tokenidx=0;}
			else if (state!=6 && tokenidx<63)
			{
				token[state][tokenidx++]=c;
				token[state][tokenidx]=0;	
			}
		}
	}
	//printf("%s;%s;%s;%s;%s;%s\n",token[0],token[1],token[2],token[3],token[4],token[5]);
	// token0: contains the color group
	// token1: contains fg or bg
	// token2: contains the color
	// token3: contains fg or bg
	// token4: contains the color black, red, green, blue, purple, cyan, brown, yellow, white, lightblack, lightred, lightgreen, lightblue, lightpurple,lightcyan, lightbrown, lightyellow
	// token5: contains something light BOLD, reverse, blink, underline, dim
	
	col1=0;col2=0;
	attrs1=0;attrs2=0;
	attrsret=0;
	for (i=0;i<NUMCOLORNAMES;i++)
	{
		//printf("   colorNames[i].name=%s,token[2]=%s\n",colorNames[i].name,token[2]);
		if (strncmp(token[2],colorNames[i].name,colorNames[i].len)==0) {col1=colorNames[i].val;attrs1=colorNames[i].attrs;}
		if (strncmp(token[4],colorNames[i].name,colorNames[i].len)==0) {col2=colorNames[i].val;attrs2=colorNames[i].attrs;}
	}
	if (strncmp(token[1],"FG",2)==0) 
	{
		fg=col1;
		attrsret|=attrs1;
	}
	if (strncmp(token[1],"BG",2)==0) 
	{
		bg=col1;
		attrsret|=(attrs1==A_BOLD)?A_BLINK:attrs1;
	}

	if (strncmp(token[3],"FG",2)==0) 
	{
		fg=col2;
		attrsret|=attrs2;
	}
	if (strncmp(token[3],"BG",2)==0) 
	{
		bg=col2;
		attrsret|=(attrs2==A_BOLD)?A_BLINK:attrs2;
	}
	for (i=0;i<NUMATTRNAMES;i++)
	{
		if (strncmp(attrNames[i].name,token[5],attrNames[i].len)==0) attrsret|=attrNames[i].attrs;
	}

	for (i=0;i<NUMCOLORGROUPS;i++)
	{
		if (strncmp(colorGroups[i].configname,token[0],colorGroups[i].len)==0)
		{
			colorpair(output,colorGroups[i].uicol,fg,bg,attrsret);
		}
	}
	return	RETOK;
}
int readconfigfile(tOutput* output,char* filename)
{
	tFptr f=fopen(filename,"rb");
	char line[512];
	unsigned char c;
	int lineidx=0;
	int	keyboardcnt;
	int n;
	if (!f) return 1;
	keyboardcnt=0;
	while (!feof(f))
	{
		n=fread(&c,sizeof(char),1,f);
		if (n && !feof(f))
		{
			if (c>='a' && c<='z') c-=32;	// make everything uppercase
			if (c==9) c=32;			// and whitespaces are all the same
			if (lineidx<511 && c!=32) 
			{
				line[lineidx++]=c;
				line[lineidx]=0;
				if (c<32)
				{
					getcolors(output,line);
					keyboardcnt+=(configkeytab((tKeyTab*)output->pKeyTab,line)==0);
					lineidx=0;
					line[0]=0;
				}
			}
		}
	}
	fclose(f);
	if (keyboardcnt!=NUM_SPECIALKEYS) return 2;	// not every spcial key was configured in the config file
	return 0;
}
